blockchain/nonce_settings.go

```diff
@@ -0,0 +1,144 @@
+var (
+	globalNonceManager *NonceSettings
+	onlyOnce           sync.Once
+)
+
+// useGlobalNonceManager for when running tests on a non-simulated network
+func useGlobalNonceManager() *NonceSettings {
+	onlyOnce.Do(func() {
+		globalNonceManager = newNonceSettings()
+		go globalNonceManager.watchInstantTransactions()
+	})
+	return globalNonceManager
+}
+
+// convenience function
+func newNonceSettings() *NonceSettings {
+	return &NonceSettings{
+		NonceMu: &sync.Mutex{},
+		Nonces:  make(map[string]uint64),
+
+		doneChan:            make(chan struct{}),
+		instantTransactions: make(map[string]map[uint64]chan struct{}),
+		instantNonces:       make(map[string]uint64),
+		registerChan:        make(chan instantTxRegistration),
+		sentChan:            make(chan string),
+	}
+}
+
+// NonceSettings is a convenient wrapper for holding nonce state
+type NonceSettings struct {
+	NonceMu *sync.Mutex
+	Nonces  map[string]uint64
+
+	// used to properly meter out instant txs on L2s
+	doneChan            chan struct{}
+	instantTransactions map[string]map[uint64]chan struct{}
+	instantNonces       map[string]uint64
+	instantNoncesMu     sync.Mutex
+	registerChan        chan instantTxRegistration
+	sentChan            chan string
+}
+
+// GetNonce keep tracking of nonces per address, add last nonce for addr if the map is empty
+func (e *EthereumClient) GetNonce(ctx context.Context, addr common.Address) (uint64, error) {
+	e.NonceSettings.NonceMu.Lock()
+	defer e.NonceSettings.NonceMu.Unlock()
+	if _, ok := e.NonceSettings.Nonces[addr.Hex()]; !ok {
+		pendingNonce, err := e.Client.PendingNonceAt(ctx, addr)
+		if err != nil {
+			return 0, err
+		}
+		e.NonceSettings.Nonces[addr.Hex()] = pendingNonce
+
+		e.NonceSettings.instantNoncesMu.Lock()
+		e.NonceSettings.instantNonces[addr.Hex()] = pendingNonce
+		e.NonceSettings.instantNoncesMu.Unlock()
+
+		return pendingNonce, nil
+	}
+	e.NonceSettings.Nonces[addr.Hex()]++
+	return e.NonceSettings.Nonces[addr.Hex()], nil
+}
+
+// PeekPendingNonce returns the current pending nonce for the address. Does not change any nonce settings state
+func (e *EthereumClient) PeekPendingNonce(addr common.Address) (uint64, error) {
+	e.NonceSettings.NonceMu.Lock()
+	defer e.NonceSettings.NonceMu.Unlock()
+	if _, ok := e.NonceSettings.Nonces[addr.Hex()]; !ok {
+		pendingNonce, err := e.Client.PendingNonceAt(context.Background(), addr)
+		if err != nil {
+			return 0, err
+		}
+		e.NonceSettings.Nonces[addr.Hex()] = pendingNonce
+	}
+	return e.NonceSettings.Nonces[addr.Hex()], nil
+}
+
+// watchInstantTransactions should only be called when minConfirmations for the chain is 0, generally an L2 chain.
+// This helps meter out transactions to L2 chains, so that nonces only send in order. For most (if not all) L2 chains,
+// the mempool is small or non-existent, meaning we can't send nonces out of order, otherwise the tx is instantly
+// rejected.
```

Comment:
Can we, possibly, remove the case where we can set minConfirmations = 0, I'm seeing more and more code and comments to handle and reason about that case, but I don't understand why such case as minConfirmations = 0 even exists, isn't 1 is the sane default for all networks, or even more?

Reply:
If we can't send `nonces` out of order, maybe we can just work in blocking mode if it's fast?

Reply:
Defaulting to `minConfirmations=1` for L2s like Optimism led to many flakes and slower test execution. It often created a race between how fast we subscribe to watch the transaction and when the block that includes the transaction is "mined." It also left us open to eternally waiting for transactions to confirm. It's possible in chains like Arbitrum to not have a block for many minutes or hours if the chain isn't active.

For using a blocking mode in all cases, I'm not fully against it, but I've no idea what the performance would look like for chains like Goerli afterwards. It probably wouldn't have much of an effect.